drivers/stepper/gpio_stepper/gpio_stepper_controller.c

```diff
@@ -0,0 +1,205 @@
+	}
+	return 0;
+}
+
+void stepper_work_step_handler(struct k_work *work)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(work);
+	struct gpio_stepper_motor_controller_data *data =
+		CONTAINER_OF(dwork, struct gpio_stepper_motor_controller_data, stepper_dwork);
+	static int32_t step_count = 1;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	LOG_DBG("Moving Motor from %d to %d with delay %d ", data->actual_position,
+		data->target_position, data->delay_in_us);
+	if (data->actual_position != data->target_position) {
+		if (data->actual_position > data->target_position) {
+			stepper_motor_move_with_4_gpios(data->control_pins, step_count--);
+			data->actual_position--;
+		} else if (data->actual_position < data->target_position) {
+			stepper_motor_move_with_4_gpios(data->control_pins, step_count++);
+			data->actual_position++;
+		}
+		k_work_reschedule(&data->stepper_dwork, K_USEC(data->delay_in_us));
+	}
+	k_spin_unlock(&data->lock, key);
+}
+
+static int32_t gpio_stepper_move(const struct device *dev, int32_t micro_steps)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	data->target_position = data->actual_position + micro_steps;
+	k_spin_unlock(&data->lock, key);
+	k_work_reschedule(&data->stepper_dwork, K_NO_WAIT);
+	return 0;
+}
+
+static int32_t gpio_stepper_set_actual_position(const struct device *dev, int32_t position)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	data->actual_position = position;
+	k_spin_unlock(&data->lock, key);
+
+	k_work_reschedule(&data->stepper_dwork, K_NO_WAIT);
+	return 0;
+}
+
+static int32_t gpio_stepper_get_actual_position(const struct device *dev, int32_t *position)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+	*position = data->actual_position;
+	k_spin_unlock(&data->lock, key);
+
+	return 0;
+}
+
+static int32_t gpio_stepper_set_target_position(const struct device *dev, int32_t position)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	data->target_position = position;
+	k_spin_unlock(&data->lock, key);
+	k_work_reschedule(&data->stepper_dwork, K_NO_WAIT);
+	return 0;
+}
+
+static int gpio_stepper_is_moving(const struct device *dev, bool *is_moving)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+	*is_moving = k_work_delayable_is_pending(&data->stepper_dwork);
+	LOG_DBG("Is Motor Moving %d", *is_moving);
+	return 0;
+}
+
+static int gpio_stepper_set_max_velocity(const struct device *dev, uint32_t velocity)
+{
+	if (velocity == 0) {
+		LOG_ERR("Velocity cannot be zero");
+		return -EINVAL;
+	}
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	data->delay_in_us = (USEC_PER_SEC * SEC_PER_MIN) / velocity;
```

Comment:
Could you expand on this formula? The velocity is in steps per second right?  This formula gives you steps per minute (velocity = 1 => delay_in_us = 60000000) which is 1 step per minute